File: skiena_1_6_2.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

struct field_t
{
  // data has 2 more lines and 2 more columns than the actual number of
  // lines and columns read from input file. The extra lines and columns will
  // be initialized to '+'. This serves to facilitate calculating the number
  // of mines in the surrounding cells of any particular cell. So, if the
  // input_t from input file is:
  //		*...
  //		....
  //		.*..
  //		....
  // Then the internal data input_t looks something like this:
  //		++++++
  //		+*...+
  //		+....+
  //		+.*..+
  //		++++++
  char** data;
  int row_count;
  int column_count;
};
typedef struct field_t field_t;

// this method creates a new string having the desired length, and padd the
// contents of such string with blank characters "+"
char* blanks(int size)
{
  char* line = malloc(size);
  memset(line, '+', size-1);
  line[size-1] = '\0';
  return line;
}

// this constructor initializes internal data from a list of strings
#define MAX_ROW 100
#define MAX_COLUMN 100
void fi_init_input(field_t* field, int row, int column, char list[][MAX_COLUMN])
{
// ("inside fi_init_input() field: %p\n", field);
  field->row_count = row + 2;
  field->column_count = column + 2;
  field->data = malloc(field->row_count * sizeof(char*));
  // add a blank string at the beginning
  field->data[0] = blanks(column + 3);

  // store the date while padding a blank at both the beginning and the end
  int i;
  for (i = 0; i < row; i++)
  {
    field->data[i+1] = malloc(column + 3);
    field->data[i+1][0] = '+';
    strcpy(field->data[i+1]+1, list[i]);
    strcat(field->data[i+1], "+");
  }

  // add a blank string to the end
  field->data[row+1] = blanks(column + 3);
}

// this method builds the internal data field, by recording the mines where
// they are present in data, and also the number of mines in the surrounding
// 8 cells, for each cell
void fi_init_output(field_t* output, field_t* input)
{
  // remember: input has 2 extra lines
  output->row_count = input->row_count - 2;
  output->data = malloc(output->row_count * sizeof(char*));

  int i;
  for (i = 1; i < input->row_count-1; i++)
  {
    char* builder = malloc(input->column_count-1);
    builder[0] = '\0';
    // each line in data has 2 extra columns (at beginning and at end)
    int j;
    for (j = 1; j < input->column_count-1; j++)
    {
      // if this input cell contains a mine, record the mine in the output cell
      if (input->data[i][j] == '*')
        strcat(builder, "*");
      // otherwise, calculate the number of mines in the surrounding 8 cells
      // and record that number in the output cell
      else
        sprintf(builder+strlen(builder), "%c", ('0' + fi_mine_total(input, i, j)));
    }
    // retain the string
    output->data[i-1] = builder;
  }
}

void fi_destructor(field_t* field)
{
  int i;
  for (i = 0; i < field->row_count; i++)
    free(field->data[i]);
  free(field->data);
}

// this method prints the original field as read from STDIN
char* fi_tostring_original(field_t* field, char* string)
{
  string[0] = '\0';
  int i;
  for (i = 1; i < field->row_count-1; i++)
  {
    char tmp[MAX_COLUMN];
    strncpy(tmp, field->data[i]+1, field->column_count-2);
    tmp[field->column_count-2] = '\n';
    tmp[field->column_count-1] = '\0';
    strcat(string, tmp);
  }
  return string;
}

// this method prints everything in data
char* fi_tostring(field_t* field, char* string)
{
  string[0] = '\0';
  int i;
  for (i = 0; i < field->row_count; i++)
    sprintf(string+strlen(string), "%s\n", field->data[i]);
  return string;
}

// this method returns whether this cell contain a mine
int mined(char value)
{
  return (value == '*') ? 1 : 0;
}

// this method returns the total number of mines in the surrounding 8 cells
int fi_mine_total(field_t* field, int i, int j)
{
  return mined(field->data[i-1][j-1]) +
         mined(field->data[i-1][j]) +
         mined(field->data[i-1][j+1]) +
         mined(field->data[i][j-1]) +
         mined(field->data[i][j+1]) +
         mined(field->data[i+1][j-1]) +
         mined(field->data[i+1][j]) +
         mined(field->data[i+1][j+1]);
}

// this method reads data from STDIN. The first line of each input_t contains 2
// integers which stand for the number of lines and columns of the input_t. Each
// of the next n lines contains exactly m characters, representing the input_t.
// Safe squares are denoted by "." and mine squares by "*". The first input_t
// line where n = m = 0 represents the end of input and should not be processed.
void input(field_t list[], int* size)
{
  *size = 0;
  while (1)
  {
    // read 2 integers
    int line_count, column_count;
    scanf("%d%d", &line_count, &column_count);
    // throw away the EOL, to prepare for the next line of text
    char tmp[20];
    fgets(tmp, 20, stdin);

    // end of input
    if (line_count == 0 || column_count == 0)
      break;
    else
    {
      char lines[MAX_ROW][MAX_COLUMN];
      // read the next line_count lines
      int i;
      for (i = 0; i < line_count; i++)
      {
        // allocate 1 extra character to accommodate the \n character from fgets()
        fgets(lines[i], column_count+2, stdin);
        // throw away the \n character
        lines[i][column_count] = '\0';
      }

      // create and retain a field of lines
      fi_init_input(list + *size, line_count, column_count, lines);
      *size += 1;
    }
  }
}

// this method sweeps thru a list of fields; for each field it prints out the
// mine cells as well as cells which contain the total number of mines in
// the surrounding 8 cells
void output(field_t list[], int size)
{
  // use traditional for loop instead of the advanced for loop because index
  // is needed in printing
  int i;
  for (i = 0; i < size; i++)
  {
    // for each cell in the input_t, sweep the surrounding cells for mines and
    // store the number of mines in the cell
    field_t out;
    fi_init_output(&out, list+i);

    // print output in the format required
    printf("\n");
    printf("Field #%d:\n", i+1);
    char string[MAX_ROW*MAX_COLUMN];
    printf("%s", fi_tostring(&out, string));

    fi_destructor(list+i);
    fi_destructor(&out);
  }
}

// assume for now maximum of 100 test fields
#define MAX_FIELD 100
void main()
{
  field_t list[MAX_FIELD];
  int size;
  input(list, &size);
  output(list, size);
}

File: skiena_1_6_2.rb
# this class is a base class for the 2 classes PaddedField and CalculatedField.
# it declares a common data field, and a common toString method.
class Field
  def to_s
    string = ""
    @data.each do |row|
      string << row << "\n"
    end
    string
  end
end

# this class contains the data field as read from input, with 2 extra lines and
# 2 extra columns added as 4 borders around the actual data field. the borders
# serve to facilitate calculating the number of mines in the 8 surrounding
# cells of any cell. So, if the field from input file is:
#		*...
#		....
#		.*..
#		....
# Then the internal data field looks like this:
#		++++++
#		+*...+
#		+....+
#		+.*..+
#		++++++
class PaddedField < Field
  attr_accessor :data

  # this constructor initializes internal data from a list of strings
  def initialize(list)
    @data = []

    # add a blank string at the beginning
    @data << blanks(list[0].length + 2)

    # store the date while padding a blank at both the beginning and the end
    list.each do |string|
      tmp = ""
      tmp << "+" << string << "+"
      @data << tmp
    end

    # add a blank string to the end
    @data << blanks(list[0].length + 2)
  end

  # this method returns the total number of mines in the surrounding 8 cells
  def mine_total(i, j)
    return is_mined(@data[i-1][j-1]) +
           is_mined(@data[i-1][j]) +
           is_mined(@data[i-1][j+1]) +
           is_mined(@data[i][j-1]) +
           is_mined(@data[i][j+1]) +
           is_mined(@data[i+1][j-1]) +
           is_mined(@data[i+1][j]) +
           is_mined(@data[i+1][j+1]);
  end

private
  # this method creates a new string having the desired length, and padd the
  # contents of such string with blank characters "+"
  def blanks(size)
    "+" * size
  end

  # this method prints the data field exactly as read from input (with the 4
  # borders omitted)
  def to_s2()
    builder = ""
    (1..(data.size-2)).each do |i|
      (1..(data[i].size-2)).each do |j|
        builder << data[i][j]
      end
      builder << '\n'
    end
    builder
  end

  # this method returns whether this cell contain a mine
  def is_mined(value)
    (value == '*') ? 1 : 0
  end
end

# this class represents the field whose every cell contains the calculated
# number of mines in its surrounding 8 cells.
class CalculatedField < Field
  # this method builds the internal data input_t, by recording the mines where
  # they are present in data, and also the number of mines in the surrounding
  # 8 cells, for each cell
  def initialize(input)
    @data = []
    # remember: data has 2 extra lines
    1.upto(input.data.length-2) do |i|
      builder = ""
      # each line in data has 2 extra columns (at beginning and at end)
      1.upto(input.data[i].length-2) do |j|
        # if this input cell contains a mine, record the mine in the output cell
        if (input.data[i][j] == '*')
          builder << "*"
        # otherwise, calculate the number of mines in the surrounding 8 cells
        # and record that number in the output cell
        else
          builder << ('0'.ord + input.mine_total(i, j)).chr
        end
      end
      # retain the string
      @data << builder
    end
  end
end

# this method reads data from STDIN. The first line of each input_t contains 2
# integers which stand for the number of lines and columns of the input_t. Each
# of the next n lines contains exactly m characters, representing the input_t.
# Safe squares are denoted by "." and mine squares by "*". The first input_t
# line where n = m = 0 represents the end of input and should not be processed.
def input
  list = []
  while (true)
    # read 2 integers
    # also throw away the EOL, to prepare for the next line of text
    counts = gets.split.map(&:to_i)
    line_count = counts[0]
    column_count = counts[1]

    # end of input
    if (line_count == 0 || column_count == 0)
      break;
    else
      lines = []
      # read the next line_count lines
      line_count.times do
        lines << gets.chomp!
      end

      # create and retain a input_t of lines
      list << PaddedField.new(lines)
    end
  end

  list
end

# this method sweeps thru a list of input_ts; for each input_t it prints out the
# mine input_ts as well as input_ts which contain the total number of mines in
# the surrounding 8 cells
def output(list)
  # use traditional for loop instead of the advanced for loop because index
  # is needed in printing
  list.size.times do |i|
    # for each cell in the input_t, sweep the surrounding cells for mines and
    # store the number of mines in the cell
    output = CalculatedField.new(list[i])

    # print output in the format required
    puts
    puts("Field ##{i+1}:")
    puts output
  end
end

output(input())
